File: src/TurboVision/Views/WriteArgs.cs
using System.Runtime.InteropServices;

namespace TurboVision.Views;

public unsafe partial class WriteArgs(void* ptr, bool owned, bool placement)
    : NativeObject<WriteArgs>(ptr, owned, placement)
{
    private sealed class Factory : NativeObjectFactory<Factory>
    {
        public Factory()
            : base(
                NativeMethods.TV_write_args_placementSize,
                NativeMethods.TV_write_args_placementNew,
                NativeMethods.TV_write_args_new
            ) { }
    }

    public static int PlacementSize => Factory.Instance.PlacementSize;

    public WriteArgs(byte* placement)
        : this(Factory.Instance.PlacementNew(placement), owned: true, placement: true) { }

    public WriteArgs()
        : this(Factory.Instance.New(), owned: true, placement: false) { }

    protected override void PlacementDeleteCore(void* ptr)
    {
        TurboVisionException.Check(NativeMethods.TV_write_args_placementDelete(ptr));
    }

    protected override void DeleteCore()
    {
        TurboVisionException.Check(NativeMethods.TV_write_args_delete(Ptr));
    }

    protected override bool EqualsCore(WriteArgs other)
    {
        TurboVisionException.Check(
            NativeMethods.TV_write_args_equals(Ptr, other.Ptr, out var equals)
        );
        return equals;
    }

    protected override int GetHashCodeCore()
    {
        TurboVisionException.Check(NativeMethods.TV_write_args_hash(Ptr, out var hash));
        return hash;
    }

    public void* Self
    {
        get
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            TurboVisionException.Check(NativeMethods.TV_write_args_get_self(Ptr, out var self));
            return self;
        }
        set
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            TurboVisionException.Check(NativeMethods.TV_write_args_set_self(Ptr, value));
        }
    }

    public void* Target
    {
        get
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            TurboVisionException.Check(NativeMethods.TV_write_args_get_target(Ptr, out var target));
            return target;
        }
        set
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            TurboVisionException.Check(NativeMethods.TV_write_args_set_target(Ptr, value));
        }
    }

    public void* Buf
    {
        get
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            TurboVisionException.Check(NativeMethods.TV_write_args_get_buf(Ptr, out var buf));
            return buf;
        }
        set
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            TurboVisionException.Check(NativeMethods.TV_write_args_set_buf(Ptr, value));
        }
    }

    public ushort Offset
    {
        get
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            TurboVisionException.Check(NativeMethods.TV_write_args_get_offset(Ptr, out var offset));
            return offset;
        }
        set
        {
            ObjectDisposedException.ThrowIf(IsDisposed, this);
            TurboVisionException.Check(NativeMethods.TV_write_args_set_offset(Ptr, value));
        }
    }

    internal static partial class NativeMethods
    {
        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_write_args_placementSize(
            out int outSize,
            out int outAlignment
        );

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_write_args_placementNew(byte* self);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_write_args_placementDelete(void* self);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_write_args_new(out void* @out);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_write_args_delete(void* self);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_write_args_equals(
            void* self,
            void* other,
            [MarshalAs(UnmanagedType.I4)] out bool @out
        );

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_write_args_hash(void* self, out int @out);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_write_args_get_self(void* self, out void* @out);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_write_args_set_self(void* self, void* value);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_write_args_get_target(void* self, out void* @out);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_write_args_set_target(void* self, void* value);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_write_args_get_buf(void* self, out void* @out);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_write_args_set_buf(void* self, void* value);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_write_args_get_offset(void* self, out ushort @out);

        [LibraryImport(Global.DLL_NAME)]
        public static partial Error TV_write_args_set_offset(void* self, ushort value);
    }
}

File: src/tvision4c/Views/WriteArgs.cpp
#include "WriteArgs.h"

EXPORT tv::Error TV_WriteArgs_placementSize(int32_t* outSize, int32_t* outAlignment) {
    return tv::checkedSize<write_args>(outSize, outAlignment);
}

EXPORT tv::Error TV_WriteArgs_placementNew(write_args* self) {
    return tv::checkedPlacementNew(self);
}

EXPORT tv::Error TV_WriteArgs_placementDelete(write_args* self) {
    return tv::checkedPlacementDelete(self);
}

EXPORT tv::Error TV_WriteArgs_new(write_args** out) {
    return tv::checkedNew<write_args>(out);
}

EXPORT tv::Error TV_WriteArgs_delete(write_args* self) {
    return tv::checkedDelete(self);
}

EXPORT tv::Error TV_WriteArgs_equals(write_args* self, write_args* other, BOOL* out) {
    if (!out) {
        return tv::Error_ArgumentNull;
    }

    if (!self && !other) {
        *out = TRUE;
        return tv::Success;
    }

    if (!self || !other) {
        *out = FALSE;
        return tv::Success;
    }

    if (self->self == other->self && self->target == other->target && self->buf == other->buf &&
        self->offset == other->offset) {
        *out = TRUE;
        return tv::Success;
    }

    *out = FALSE;
    return tv::Success;
}

EXPORT tv::Error TV_WriteArgs_hash(write_args* self, int32_t* out) {
    if (!self || !out) {
        return tv::Error_ArgumentNull;
    }

    int32_t result = 0;
    tv::hash(reinterpret_cast<uintptr_t>(self->self), &result);
    tv::hash(reinterpret_cast<uintptr_t>(self->target), &result);
    tv::hash(reinterpret_cast<uintptr_t>(self->buf), &result);
    tv::hash(self->offset, &result);

    *out = result;
    return tv::Success;
}

EXPORT tv::Error TV_WriteArgs_get_self(write_args* self, void** out) {
    if (!self || !out) {
        return tv::Error_ArgumentNull;
    }

    *out = self->self;
    return tv::Success;
}

EXPORT tv::Error TV_WriteArgs_set_self(write_args* self, void* value) {
    if (!self) {
        return tv::Error_ArgumentNull;
    }

    self->self = value;
    return tv::Success;
}

EXPORT tv::Error TV_WriteArgs_get_target(write_args* self, void** out) {
    if (!self || !out) {
        return tv::Error_ArgumentNull;
    }

    *out = self->target;
    return tv::Success;
}

EXPORT tv::Error TV_WriteArgs_set_target(write_args* self, void* value) {
    if (!self) {
        return tv::Error_ArgumentNull;
    }

    self->target = value;
    return tv::Success;
}

EXPORT tv::Error TV_WriteArgs_get_buf(write_args* self, void** out) {
    if (!self || !out) {
        return tv::Error_ArgumentNull;
    }

    *out = self->buf;
    return tv::Success;
}

EXPORT tv::Error TV_WriteArgs_set_buf(write_args* self, void* value) {
    if (!self) {
        return tv::Error_ArgumentNull;
    }

    self->buf = value;
    return tv::Success;
}

EXPORT tv::Error TV_WriteArgs_get_offset(write_args* self, uint16_t* out) {
    if (!self || !out) {
        return tv::Error_ArgumentNull;
    }

    *out = self->offset;
    return tv::Success;
}

EXPORT tv::Error TV_WriteArgs_set_offset(write_args* self, uint16_t value) {
    if (!self) {
        return tv::Error_ArgumentNull;
    }

    self->offset = value;
    return tv::Success;
}
